test/powershell/engine/Api/TypeInference.Tests.ps1

```diff
@@ -1494,6 +1518,178 @@ Describe "Type inference Tests" -tags "CI" {
         $null = [AstTypeInference]::InferTypeOf($FoundAst)
     }
 
+    It 'Ignores type constraint defined outside of scope' {
+        $res = [AstTypeInference]::InferTypeOf(({
+            function Outer
+            {
+                [string] $Test = "Hello"
+                function Inner
+                {
+                    $Test = 2
+                    $Test
+                }
+            }
+        }.Ast.FindAll({param($Ast) $Ast -is [Language.VariableExpressionAst]}, $true) | Select-Object -Last 1 ))
+        $res.Name | Should -Be 'System.Int32'
+    }
```

Comment:
Will the inference work after follow swap?
```c#
    It 'Ignores type constraint defined outside of scope' {
        $res = [AstTypeInference]::InferTypeOf(({
            function Outer
            {
                function Inner
                {
                    $Test = 2
                    $Test
                }
                [string] $Test = "Hello"
                Inner
            }
        ...
        $res.Name | Should -Be 'System.Int32'
    }
```

Reply:
Yes, `$Test` inside the `Inner` function will also be inferred as an int32 in this scenario because there's an assignment inside `Inner` so outer scope type constraints/assignments don't matter.

Reply:
And if we remove ` $Test = 2` the inference will stop?

Reply:
Yes. It would stop the search before it would reach the assignment because it's after the start of the `Inner` scope. This is not accurate to how PowerShell works, but trying to keep track of the variable values before each command invocation is quite difficult.

Reply:
Thanks! For idea, we would need a `safe interpreter`.